File: typescript/agent.ts
/**
 * Order fulfillment agent — TypeScript example.
 *
 * Validates inventory, charges payment, ships. In Temporal you'd need
 * 3 Activities + 1 Workflow + 1 Worker. Here it's one agent.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "order-fulfillment-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const orderId = payload.order_id ?? "unknown";
  const items: any[] = payload.items ?? [];
  const customer = payload.customer_id ?? "unknown";

  // Step 1: Validate inventory
  console.log(`  [1/3] Validating inventory for ${items.length} item(s)...`);
  await new Promise((r) => setTimeout(r, 1000));

  // Step 2: Charge payment
  const total = items.reduce(
    (sum: number, item: any) => sum + (item.price ?? 0) * (item.quantity ?? 1),
    0
  );
  console.log(`  [2/3] Charging $${total.toFixed(2)} for customer ${customer}...`);
  await new Promise((r) => setTimeout(r, 1000));

  // Step 3: Create shipment
  console.log(`  [3/3] Creating shipment for order ${orderId}...`);
  await new Promise((r) => setTimeout(r, 1000));

  const result = {
    action: "complete",
    order_id: orderId,
    total_charged: total,
    tracking_number: "TRK-98765",
    shipped_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "order-fulfillment-demo" });
  console.log(`  Order ${orderId} fulfilled. Tracking: ${result.tracking_number}`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;
    if (!intentId) continue;
    if (["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: java/Agent.java
/*
 * Order fulfillment agent — Java example.
 *
 * Fetches an intent by ID, processes order (inventory, payment, shipping),
 * and resumes with result.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   javac -cp axme-sdk.jar Agent.java
 *   java -cp .:axme-sdk.jar Agent <intent_id>
 */

import dev.axme.sdk.AxmeClient;
import dev.axme.sdk.AxmeClientConfig;
import dev.axme.sdk.RequestOptions;
import java.time.Instant;
import java.util.List;
import java.util.Map;

public class Agent {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Agent <intent_id>");
            System.exit(1);
        }

        String apiKey = System.getenv("AXME_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("Error: AXME_API_KEY not set.");
            System.exit(1);
        }

        String intentId = args[0];
        var client = new AxmeClient(AxmeClientConfig.forCloud(apiKey));

        System.out.println("Processing intent: " + intentId);

        var intentData = client.getIntent(intentId, new RequestOptions());
        @SuppressWarnings("unchecked")
        var intent = (Map<String, Object>) intentData.getOrDefault("intent", intentData);
        @SuppressWarnings("unchecked")
        var payload = (Map<String, Object>) intent.getOrDefault("payload", Map.of());
        if (payload.containsKey("parent_payload")) {
            @SuppressWarnings("unchecked")
            var pp = (Map<String, Object>) payload.get("parent_payload");
            payload = pp;
        }

        String orderId = (String) payload.getOrDefault("order_id", "unknown");
        @SuppressWarnings("unchecked")
        var items = (List<Map<String, Object>>) payload.getOrDefault("items", List.of());
        String customer = (String) payload.getOrDefault("customer_id", "unknown");

        // Step 1: Validate inventory
        System.out.println("  [1/3] Validating inventory for " + items.size() + " item(s)...");
        Thread.sleep(1000);

        // Step 2: Charge payment
        double total = 0;
        for (var item : items) {
            double price = item.containsKey("price") ? ((Number) item.get("price")).doubleValue() : 0;
            double qty = item.containsKey("quantity") ? ((Number) item.get("quantity")).doubleValue() : 1;
            total += price * qty;
        }
        System.out.printf("  [2/3] Charging $%.2f for customer %s...%n", total, customer);
        Thread.sleep(1000);

        // Step 3: Create shipment
        System.out.println("  [3/3] Creating shipment for order " + orderId + "...");
        Thread.sleep(1000);

        var result = Map.<String, Object>of(
            "action", "complete",
            "order_id", orderId,
            "total_charged", total,
            "tracking_number", "TRK-98765",
            "shipped_at", Instant.now().toString()
        );

        client.resumeIntent(intentId, result, new RequestOptions());
        System.out.println("  Order " + orderId + " fulfilled. Tracking: TRK-98765");
    }
}
